packages/lodestar/src/eth1/impl/ethers.ts

```diff
@@ -226,26 +173,122 @@ export class EthersEth1Notifier extends (EventEmitter as { new(): Eth1EventEmitt
     const address = this.opts.depositContract.address;
     const abi = this.opts.depositContract.abi;
     if (!(await this.contractExists(address))) {
-      throw new Error(`There is no deposit contract at given address: ${address}`);
+      throw new Error(`eth1: There is no deposit contract at given address: ${address}`);
     }
     try {
       this.contract = new ethers.Contract(address, abi, this.provider);
     } catch (e) {
-      throw new Error("Eth1 deposit contract not found! Probably wrong eth1 rpc url");
+      throw new Error("eth1 deposit contract not found! Probably wrong eth1 rpc url");
+    }
+  }
+
+  private async doGetDepositEventsFromLastProcessedBlock(): Promise<void> {
+    const followBlockNumber = await this.provider.getBlockNumber() - this.config.params.ETH1_FOLLOW_DISTANCE;
+    this.logger.info(
+      `eth1: Start scanning Eth1Data last block =${this.lastProcessedEth1BlockNumber}` +
+      ` followBlockNumber=${followBlockNumber}`
+    );
+    await this.process(followBlockNumber);
+    if (this.lastProcessedEth1BlockNumber === followBlockNumber) {
+      // the previous pipe is done, need to start a new one
+      this.process(undefined);
+      this.provider.on("block", this.onNewEth1Block.bind(this));
+    }
+  }
+
+
+  /**
+   * Scan deposit events per every n (100 currently) blocks until inBlockNumber
+   * If undefined inBlockNumber then this is a reusable pipe, process whenever we push to targetBlockSource.
+   * If inBlockNumber exists, we close pipe after use.
+   * @param inBlockNumber
+   */
+  private async process(inBlockNumber?: number): Promise<void> {
```

Comment:
I feel like this and methods it's using are way too complicated and have way too much responsibilities.

Reply:
yeah really, I thought you want to use `pipe` in one of your comments. I can come back to the old way
```
while (this.lastProcessedEth1BlockNumber < followUp && (this.isStarted || this.isScanForGenesis)) {
  awai this.getAndProcessDepositEvents(this.lastProcessedEth1BlockNumber + 1, blockNumber);
}
```

Reply:
It's not pipe that's problematic, problem is that this method serves to fetch deposit events from single block (which should be separate method "getBlockDepositEvents("block")) and also does batch processing which feeds events into another "source" which would make more sense to be getBatchEth1EventSource where pipe would return AsyncGenerator<DepositEvent> or something

Reply:
no, we always get deposit events in batch here. Even in the case of new block, we get deposit events from last processed block up to the follow up block to make sure we don't miss any events of blocks.

So the only business here is to simply get events up to a block and feed it into another source.